FV3/wrapper/tests/test_wind_transformations.py

```diff
@@ -0,0 +1,121 @@
+            [
+                "x_wind",
+                "y_wind",
+                "eastward_wind_before_physics",
+                "northward_wind_before_physics",
+                "eastward_wind_after_physics",
+                "northward_wind_after_physics",
+            ]
+        )
+        x_wind_before_physics = pre_applied_physics_state["x_wind"]
+        y_wind_before_physics = pre_applied_physics_state["y_wind"]
+
+        u_before_physics = pre_applied_physics_state["eastward_wind_before_physics"]
+        u_after_physics = pre_applied_physics_state["eastward_wind_after_physics"]
+
+        v_before_physics = pre_applied_physics_state["northward_wind_before_physics"]
+        v_after_physics = pre_applied_physics_state["northward_wind_after_physics"]
+
+        # Note that 3D variables from the physics component of the model have
+        # their vertical dimension flipped with respect to 3D variables from the
+        # dynamical core.  Therefore we need to flip the vertical dimension of
+        # these A-grid wind increments so that when we convert them to D-grid
+        # wind increments, they align with the D-grid winds in the dynamical
+        # core.
+        #
+        # deepcopy calls here are used out of convenience to construct Quantity
+        # objects of the same shape and metadata as others.  Their data is
+        # overwritten immediately.
+        u_increment = deepcopy(u_after_physics)
+        u_increment.view[:] = u_after_physics.view[:] - u_before_physics.view[:]
+        u_increment.view[:] = u_increment.view[:][::-1, :, :]
+
+        v_increment = deepcopy(v_after_physics)
+        v_increment.view[:] = v_after_physics.view[:] - v_before_physics.view[:]
+        v_increment.view[:] = v_increment.view[:][::-1, :, :]
+
+        (
+            x_wind_physics_increment,
+            y_wind_physics_increment,
+        ) = fv3gfs.wrapper.transform_agrid_winds_to_dgrid_winds(
+            u_increment, v_increment
+        )
+
+        fv3gfs.wrapper.apply_physics()
+        updated_dynamical_core_state = fv3gfs.wrapper.get_state(["x_wind", "y_wind"])
+        x_wind_after_physics = updated_dynamical_core_state["x_wind"]
+        y_wind_after_physics = updated_dynamical_core_state["y_wind"]
+
+        np.testing.assert_allclose(
+            x_wind_before_physics.view[:] + x_wind_physics_increment.view[:],
+            x_wind_after_physics.view[:],
+        )
+        np.testing.assert_allclose(
+            y_wind_before_physics.view[:] + y_wind_physics_increment.view[:],
+            y_wind_after_physics.view[:],
+        )
+
+
+if __name__ == "__main__":
+    config = get_default_config()
+    # Deactivate fv_sg_adj for these tests. fv_sg_adj can otherwise introduce an
+    # additional momentum tendency on the A-grid winds that is not accounted for
+    # by the difference between the A-grid winds at the start of the physics and
+    # the A-grid winds at the end of the physics.  This is just to make the
```

Comment:
Is this because the A-grid winds are not consistent with the 'true' dynamical core D-grid winds? Or some other reason? I'm not sure I follow.

Reply:
The reason is a bit in the weeds; I can try to improve this comment.  Note that [`u_dt` and `v_dt` are global variables](https://github.com/ai2cm/fv3gfs-fortran/blob/2a3ea739723d6152fd03d419c76d25741f7fa9da/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90#L275) in the `atmosphere.F90` module, so their state is carried between subroutines.  It happens that they are [reset to zero towards the end of the dynamics subroutine](https://github.com/ai2cm/fv3gfs-fortran/blob/2a3ea739723d6152fd03d419c76d25741f7fa9da/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90#L840-L842) just before the call to `fv_subgrid_z`.  [`fv_subgrid_z` (if active) then modifies them](https://github.com/ai2cm/fv3gfs-fortran/blob/2a3ea739723d6152fd03d419c76d25741f7fa9da/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90#L860-L866), and [this state is carried over into the physics update](https://github.com/ai2cm/fv3gfs-fortran/blob/2a3ea739723d6152fd03d419c76d25741f7fa9da/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90#L1618-L1619).

My reason for turning `fv_subgrid_z` off is just so that I don't need to worry about also recording and including the `fv_subgrid_z` increment in the A-grid to D-grid test.  In principle if we did record it the test would still pass and we wouldn't have to worry about turning it off.

Reply:
Okay thanks. I agree turning off for the purposes of this test is fine. Wanted to make sure this is not biting us elsewhere.

Reply:
An implication (independent of this PR) is that the `fv_subgrid_z` tendencies are included in the bulk physics tendency diagnostics that we output.  I don't think this is necessarily a concern, but something to be aware of.

Reply:
Yeah, I was wondering about this, more with respect to the budget output from SHiELD.

Reply:
Upon closer inspection it looks like `fv_subgrid_z` does not modify `t_dt` or `q_dt` at all (despite taking them as arguments), which I guess is consistent with the fact that I was able to approximately close the per-physics-component diagnostics budgets way back when without worrying about this.  *It instead modifies temperature and the tracers directly, so I believe those tendencies would not be included in our budget as it currently stands.*

It only defers the `u_dt` and `v_dt` updates to the physics update step (perhaps understandably due to the requirement of the coordinate transformation).